File: models/bookingModel.js
const pool = require('../config/db');

const getActiveBooking = (userId) => {
    return pool.query(
        `SELECT b.*, r.room_number, r.hostel, r.price, bd.bed_number, bd.position
         FROM bookings b JOIN rooms r ON b.room_id = r.id
         LEFT JOIN beds bd ON b.bed_id = bd.id
         WHERE b.user_id = ? AND b.status IN ('pending','approved')
         ORDER BY b.created_at DESC LIMIT 1`,
        [userId]
    );
};

const getStudentBookings = (userId) => {
    return pool.query(
        `SELECT b.*, r.room_number, r.hostel, r.campus, r.price, r.room_type,
         bd.bed_number, bd.position AS bed_position
         FROM bookings b JOIN rooms r ON b.room_id = r.id
         LEFT JOIN beds bd ON b.bed_id = bd.id
         WHERE b.user_id = ? ORDER BY b.created_at DESC`,
        [userId]
    );
};

const getAllBookings = () => {
    return pool.query(
        `SELECT b.*, r.room_number, r.hostel, r.campus, r.price,
         u.full_name, u.reg_no, u.email,
         bd.bed_number, bd.position AS bed_position,
         p.status AS payment_status, p.amount AS payment_amount
         FROM bookings b
         JOIN rooms r ON b.room_id = r.id
         JOIN users u ON b.user_id = u.id
         LEFT JOIN beds bd ON b.bed_id = bd.id
         LEFT JOIN payments p ON p.booking_id = b.id
         ORDER BY b.created_at DESC`
    );
};

const checkExistingBooking = (userId) => {
    return pool.query(
        "SELECT id FROM bookings WHERE user_id = ? AND status IN ('pending','approved')",
        [userId]
    );
};

const createBooking = (userId, roomId, bedId, semester, arrivalDate) => {
    return pool.query(
        'INSERT INTO bookings (user_id, room_id, bed_id, semester, arrival_date) VALUES (?, ?, ?, ?, ?)',
        [userId, roomId, bedId || null, semester, arrivalDate || null]
    );
};

const getBookingById = (id) => {
    return pool.query(
        `SELECT b.*, r.room_number, bd.bed_number, u.full_name, u.reg_no
         FROM bookings b JOIN rooms r ON b.room_id = r.id
         LEFT JOIN beds bd ON b.bed_id = bd.id
         JOIN users u ON b.user_id = u.id
         WHERE b.id = ?`, [id]
    );
};

const getBookingByIdAndUser = (id, userId) => {
    return pool.query('SELECT * FROM bookings WHERE id = ? AND user_id = ?', [id, userId]);
};

const updateBookingStatus = (id, status) => {
    return pool.query('UPDATE bookings SET status = ? WHERE id = ?', [status, id]);
};

const suspendBooking = (id, suspended, reason) => {
    return pool.query('UPDATE bookings SET suspended = ?, suspend_reason = ? WHERE id = ?',
        [suspended, suspended ? (reason || null) : null, id]);
};

const getApprovedBookingWithDetails = (id) => {
    return pool.query(
        `SELECT b.*, r.room_number, u.full_name FROM bookings b
         JOIN rooms r ON b.room_id = r.id JOIN users u ON b.user_id = u.id
         WHERE b.id = ? AND b.status = 'approved'`, [id]
    );
};

const getApprovedCount = (roomId) => {
    return pool.query(
        "SELECT COUNT(*) AS count FROM bookings WHERE room_id = ? AND status = 'approved'",
        [roomId]
    );
};

const deleteBooking = (id) => {
    return pool.query('DELETE FROM bookings WHERE id = ?', [id]);
};

const getBookingStatus = () => {
    return pool.query("SELECT message FROM notices WHERE title = 'BOOKING_STATUS' LIMIT 1");
};

const setBookingStatus = (open, userId) => {
    return pool.query("SELECT id FROM notices WHERE title = 'BOOKING_STATUS'").then(async ([existing]) => {
        if (existing.length) {
            return pool.query("UPDATE notices SET message = ? WHERE title = 'BOOKING_STATUS'", [open ? 'open' : 'closed']);
        } else {
            return pool.query("INSERT INTO notices (title, message, created_by) VALUES ('BOOKING_STATUS', ?, ?)", [open ? 'open' : 'closed', userId]);
        }
    });
};

const getStudentHostel = (userId) => {
    return pool.query(
        `SELECT r.hostel FROM bookings b JOIN rooms r ON b.room_id = r.id
         WHERE b.user_id = ? AND b.status IN ('pending','approved')
         ORDER BY b.created_at DESC LIMIT 1`, [userId]
    );
};

module.exports = {
    getActiveBooking, getStudentBookings, getAllBookings, checkExistingBooking,
    createBooking, getBookingById, getBookingByIdAndUser, updateBookingStatus,
    suspendBooking, getApprovedBookingWithDetails, getApprovedCount, deleteBooking,
    getBookingStatus, setBookingStatus, getStudentHostel
};

File: zetech_hostel_app/lib/models/booking_model.dart
class Booking {
  final String? id;
  final String? userId;
  final String? roomId;
  final String? bedId;
  final String? semester;
  final String? arrivalDate;
  final String? status;
  final bool? suspended;
  final String? roomNumber;
  final String? hostel;
  final int? bedNumber;
  final String? bedPosition;
  final double? price;
  final String? createdAt;

  Booking({this.id, this.userId, this.roomId, this.bedId, this.semester, this.arrivalDate, this.status, this.suspended, this.roomNumber, this.hostel, this.bedNumber, this.bedPosition, this.price, this.createdAt});

  factory Booking.fromJson(Map<String, dynamic> json) => Booking(
    id: (json['_id'] ?? json['id'])?.toString(),
    userId: (json['userId'] ?? json['user_id'])?.toString(),
    roomId: (json['roomId'] ?? json['room_id'])?.toString(),
    bedId: (json['bedId'] ?? json['bed_id'])?.toString(),
    semester: json['semester']?.toString(),
    arrivalDate: (json['arrivalDate'] ?? json['arrival_date'])?.toString(),
    status: json['status']?.toString(),
    suspended: json['suspended'],
    roomNumber: (json['roomNumber'] ?? json['room_number'])?.toString(),
    hostel: json['hostel']?.toString(),
    bedNumber: json['bedNumber'] ?? json['bed_number'],
    bedPosition: (json['bedPosition'] ?? json['bed_position'] ?? json['position'])?.toString(),
    price: (json['price'] as num?)?.toDouble(),
    createdAt: (json['createdAt'] ?? json['created_at'])?.toString(),
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'userId': userId,
    'roomId': roomId,
    'bedId': bedId,
    'semester': semester,
    'arrivalDate': arrivalDate,
    'status': status,
    'suspended': suspended,
    'roomNumber': roomNumber,
    'hostel': hostel,
    'bedNumber': bedNumber,
    'bedPosition': bedPosition,
    'price': price,
    'createdAt': createdAt,
  };
}
